lib/course_api.rb

```diff
@@ -0,0 +1,153 @@
+        key_file = Rails.root.join("config", "sul-harvester.key")
+        client_cert = OpenSSL::X509::Certificate.new File.read(cert_file)
+        client_key = OpenSSL::PKey.read File.read(key_file)
+        connection = Faraday::Connection.new 'https://registry.stanford.edu', :ssl => { :client_cert => client_cert, :client_key => client_key }
+        spec_course = '/doc/courseclass/' + request_class_id
+
+        begin
+            response = connection.get spec_course
+        # Timeout error
+        rescue Faraday::TimeoutError
+            # Retry
+            begin 
+                response = connection.get spec_course
+            rescue Faraday::Error
+                errors << "After timeout, retry failed for " + request_class_id.to_s
+            end
+        # Connection failed error
+        rescue Faraday::ConnectionFailed
+            errors << "Connection failed for " + request_class_id.to_s
+        rescue Faraday::ServerError
+            errors << "Server error for " + request_class_id.to_s
+        end
+        if errors.length > 0 then puts errors.to_s end
+        {"response": response, "errors": errors}
+    end
+
+    # Get course information from API request, return sections and any errors that occurred
+    def get_course_info(request_class_id)        
+        response_info = request_course_api(request_class_id)
+        response = response_info[:response]
+        errors = response_info[:errors]
+        if errors.length > 0
+            puts "Errors that have come back to get course info"
+            puts errors.length
+        end
+        sections = []
+        if(response.status == 200)
+            sections = parse_sections(response.body, request_class_id)
+        end 
+        
+        {"sections": sections, "errors": errors}
+    end
+
+    # Return sections and instructor information for a particular course API response
+    def parse_sections(response_xml, request_class_id)
+        sections = []
+        course_response = Nokogiri::XML(response_xml)
+        # Does this section have instructors, if so return the section info and instructors sunet id and name
+        # The same course request can return cross listed courses, so we will need to get a specific id
+        course_response.xpath("//class[@id='" + request_class_id + "']//section[.//instructor]").each do |section|
+            section_id = section[:id]
+            instructors = []
+            section.xpath(".//instructor/person").each do |person|
+                sunetid = person[:sunetid]
+                name = person.text
+                instructors << {"sunet": sunetid, "name": name}
+            end
+            sections << {"sid": section_id, "instructors": instructors}
+        end
+        sections
+    end
+
+    # Based on logic https://github.com/sul-dlss/registry-harvester/blob/main/Course/src/main/java/edu/stanford/BuildTermString.java
+    def generate_term(term_id)
```

Comment:
Could we document the term_id format in the method comment? I think I get it, but...